### DISCUSSION TITLE
TestDriver & waiting for PatchByQueryOperation

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#13155] [Q&A] [answered]
When using `PatchByQueryOperation` is there a way to wait for these types of operations to finalize?

I know there is a way to wait the on the operation object itself like `operation.WaitForCompletion();` but the operation is used as a fire & forget and what is needed is verifying the result in the `RavenTestDriver`.

To mention we have `WaitForIndexing` but that doesn't help here.

**Comment by garayx** [comment]
Hi 

using `operation.WaitForCompletion();` is the right way to wait until patch operation finalize.

**Reply by mcrio** [reply]
Hi. In that case, I'd need to have some logic that depending on the environment or some configuration parameter decides whether to wait or not.
I wanted to avoid that, as the waiting is required only when writing tests with the RavenDb TestDriver. That's why we have `WaitForIndexing` with the TestDriver.

**Comment by haludi** [comment]
Hi,

You can play with Changes for operation.

class TestObj
{
    public string Id { get; set; }
    public string Prop { get; set; }
}

[Fact]
public async Task TestCase()
{
    using var store = GetDocumentStore();

    using (var session = store.OpenAsyncSession())
    {
        await session.StoreAsync(new TestObj());
        await session.SaveChangesAsync();
    }

    var changes = store.Changes().ForAllOperations();
    await changes.EnsureSubscribedNow();

    var task = new TaskCompletionSource<OperationStatusChange>();
    changes.Subscribe(change =>
    {
        if (change.State.Status == OperationStatus.InProgress)
            return;

        task.SetResult(change);
    });

    await store.Operations.SendAsync(new PatchByQueryOperation(@"
from testObjs as o
update {{
o.Prop = ""Changed""
}}"));

    await task.Task;
}

Best regards,
Igal

**Reply by mcrio** [reply]
Thanks, sounds good for my case 👍

**Reply by mcrio** [reply]
Hi, as the `store.Changes().ForAllOperations();` is no longer available with the latest DocumentStore implementation, is there any way to wait for completion of operations when we do not know the operation id.

Use case are integration tests where we test a particular endpoint which may internally execute operations, for example a `PatchByQueryOperation`. As this is part of testing the actual endpoint using the in-memory ASP test server we do not know the RavenDb operation id.

**Reply by ppekrol** [reply, marked as answer]
@mcrio `ForAllOperations` is only available when you specify for which node you want to get all the operations. You can do it like follows:

```csharp
store.Changes(databaseName, nodeTag).ForAllOperations()
```

nodeTag in a single node cluster usually is 'A', if you will leave `databaseName` null then `store.Database` will be used

**Reply by mcrio** [reply]
Great, thanks for the quick response.
